lib/puppet/provider/mount/parsed.rb

```diff
@@ -272,6 +272,51 @@ def self.mountinstances
     instances
   end
 
+  # Sort fstab entries so that
+  # a mount point is located inside another mount point appears later and
+  # a device that is located inside a mount point appears later.
+  #
+  # "Stable opportunistic insertion sort"
+  # Copy unsorted list A to sorted B by repeatedly
+  # * removing the first element from A
+  # * compare with each element in B, from first to last
+  # * insert before current element in B if smaller
+  # * insert at the end of B otherwise
+  # * finish if A is empty
+  # * repeat
+  #
+  # Stabilizes itself by inserting as late as possible.
+  #
+  # A record is "smaller" than another if
+  # * the mount point of A is a leading substring of B's mount point
+  # * the mount point of A is a leading substring of B's device (B is a bind mount)
+  #
+  # @note Does not really care about directory names and does substring
+  #   comparison only. So /media will be put before /mediastuff. This is
+  #   not necessary but will be rare and usually cause no harm, either.
+  #
+  # @api private
+  def self.order(records)
+    # insert first record first
+    result = [ records[0] ]
+    # iterate over the rest
+    records.drop(1).each do |a|
```

Comment:
Nitpick: `drop(1)` will allocate a new array.  Can we avoid doing so just for skipping the first element in the enumeration?

Insertion sort could also easily be done in-place rather than creating a resulting array as well.  We probably don't want `records` mutated (haven't looked into what we do with the array after writing out the entries), so no big deal.